src/Workspaces/Core/Portable/Workspace/Solution/RecoverableTextAndVersion.cs

```diff
@@ -15,129 +16,117 @@ namespace Microsoft.CodeAnalysis
+
+            if (sourceOrRecoverableText is RecoverableText recoverableText)
             {
-                value = TextAndVersion.Create(text, _version, _filePath, _loadDiagnostic);
-                return true;
+                if (recoverableText.TryGetValue(out var text))
+                {
+                    value = TextAndVersion.Create(text, recoverableText.Version, filePath: null, recoverableText.LoadDiagnostic);
+                    return true;
+                }
             }
-            else
+            else if (((ValueSource<TextAndVersion>)sourceOrRecoverableText).TryGetValue(out value))
             {
-                value = null!;
-                return false;
+                return true;
             }
+
+            value = null;
+            return false;
         }
 
         public bool TryGetTextVersion(out VersionStamp version)
         {
-            version = _version;
+            if (_initialSourceOrRecoverableText is ITextVersionable textVersionable)
+            {
+                return textVersionable.TryGetTextVersion(out version);
+            }
 
-            // if the TextAndVersion has not been stored yet, but it has been observed
-            // then try to get version from cached value.
-            if (version == default)
+            if (TryGetValue(out var textAndVersion))
             {
-                if (TryGetValue(out var textAndVersion))
-                {
-                    version = textAndVersion.Version;
-                }
-                else if (_initialSource is ITextVersionable textVersionable)
-                {
-                    return textVersionable.TryGetTextVersion(out version);
-                }
+                version = textAndVersion.Version;
+                return true;
             }
 
-            return version != default;
+            version = default;
+            return false;
         }
 
         public override TextAndVersion GetValue(CancellationToken cancellationToken = default)
         {
-            if (_text == null)
+            if (_initialSourceOrRecoverableText is ValueSource<TextAndVersion> source)
             {
-                using (Gate.DisposableWait(cancellationToken))
-                {
-                    if (_text == null)
-                    {
-                        return InitRecoverable(_initialSource!.GetValue(cancellationToken));
-                    }
-                }
+                // replace initial source with recovarable text, if it hasn't been replaced already:
+                Interlocked.CompareExchange(
+                    ref _initialSourceOrRecoverableText,
+                    value: new RecoverableText(source.GetValue(cancellationToken), _services),
+                    comparand: source);
             }
 
-            return TextAndVersion.Create(_text.GetValue(cancellationToken), _version, _filePath, _loadDiagnostic);
+            var recoverableText = (RecoverableText)_initialSourceOrRecoverableText;
```

Comment:
safe to just read this?  do you need some sort of barrier here?

Reply:
I believe so. Interlocked.CompareExchange is the write barrier.